### PR TITLE
Upload some artifacts for "nightly releases"

### PR DESCRIPTION
**Pull request by fingolfin**
... so that we can use them for further build steps / review them
for debugging.

Not sure if I even got the syntax right. Perhaps @ssiccha or @FriedrichRober can help finishing this.

Also, right now this mixes a bunch of things together (gap-X.Y.Z.tar.gz, gap-X.Y.Z-core.tar.gz, package-info.json, and their sha256 files) -- perhaps better to upload them as multiple separate artifacts

### CODE REVIEW

File: .github/workflows/release.yml [hunk 1]
```diff
@@ -46,6 +46,17 @@ jobs:
         run: bash dev/ci-prepare.sh
       - name: "Make archives"
         run: python -u ./dev/releases/make_archives.py
+
+      - name: "Upload artifacts"
+        if: ${{ !startsWith(github.ref, 'refs/tags/v') }}
```
Comment (fingolfin) [inline .github/workflows/release.yml:51]: Uhh, I didn't even check if this syntax is valid...

File: .github/workflows/release.yml [hunk 2]
```diff
@@ -46,6 +46,17 @@ jobs:
...
+        # Warning: the result is a single .zip file (so things are compressed twice)
+        # To keep things from exploding, we only upload a subset of all generated files
+        uses: actions/upload-artifact@v2
+        with:
+          path: |
+            tmp/gap-*.tar.gz*
+            tmp/*json*
```
Comment (fingolfin) [inline .github/workflows/release.yml:58]: Note that one can also exclude patterns (see <https://github.com/actions/upload-artifact>), which can be useful to match `gap-X.Y.Z.tar.gz` but not `gap-X.Y.Z-core.tar.gz`

**Comment by FriedrichRober** [comment]
Ich schaue es mir mal an.

File: .github/workflows/release.yml (hunk 1 above)
Comment (FriedrichRober) [reply .github/workflows/release.yml:51]: The syntax should be correct according to GitHub Docs.

File: .github/workflows/release.yml (hunk 2 above)
Comment (FriedrichRober) [reply .github/workflows/release.yml:58]: I think adding `!tmp/gap-*-core.tar.gz*` should do the job to exclude `gap-X.Y.Z-core.tar.gz`

**Comment by FriedrichRober** [comment]
I am running the workflow on my fork and test if the artifacts upload correctly.

File: .github/workflows/release.yml (hunk 2 above)
Comment (FriedrichRober) [reply .github/workflows/release.yml:58]:
```suggestion
            tmp/*json*
            !tmp/gap-*-core.tar.gz*
```

**Comment by FriedrichRober** [comment]
So everything seems to work fine on my fork: [Upload-Artifacts](https://github.com/FriedrichRober/gap/actions/runs/614764414)

Changes made on my fork:
I added the suggested line for excluding the file `gap-X.Y.Z-core.tar.gz`.
I commented out line 23 in `relase.yml` so that the workflow could trigger on my branch.

**Comment by FriedrichRober** [comment]
I looked into the following [Issue](https://github.com/actions/upload-artifact/issues/3) to see if we can upload artifacts separately. As far as I understand this is currently not possible. The files are uploaded separately but at the moment we can access them through the GitHub UI only as a compressed zip file all at once.

**Comment by fingolfin** [comment]
Thanks @FriedrichRober

**Comment by fingolfin** [comment]
I am wary about merging this as-is, as it would add hundreds of MB of data *to every single build*; so if a PR is rebuilt 5 times before it is merged, you get 5 such artifacts, and they are kept by default for 90 days. So if there is a lot of activity, the data accumulates quickly... At least that's how I understand it to  work?

Maybe more reasonable is to keep the archives only if a special flag is set. I.e., if a PR *fails* the release test, then we want to be able to get a look at the archives; perhaps this could be done by checking for a specific label on the PR? Not sure if (and how) we could access such information, though. Another alternative would be to add a manual trigger for the "release" workflow, and *that* uploads archives.

Finally, we still want to set up release builds that happen on a clock (say 24 hours), and for those we also will want to upload the archives?

**Comment by ssiccha** [comment]
@fingolfin 
> Maybe more reasonable is to keep the archives only if a special flag is set. I.e., if a PR fails the release test

Ha, I've actually found what you're looking for a few days ago! We can do [this](https://docs.github.com/en/actions/reference/context-and-expression-syntax-for-github-actions#failure) to run a step only if the job itself failed.

**Comment by ssiccha** [comment]
See the commit that I just pushed.

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -48,7 +48,7 @@ jobs:
         run: python -u ./dev/releases/make_archives.py
 
       - name: "Upload artifacts"
-        if: ${{ !startsWith(github.ref, 'refs/tags/v') }}
+        if: ${{ !startsWith(github.ref, 'refs/tags/v') && failure() }}
```
Comment (fingolfin) [inline .github/workflows/release.yml:51]: nice!

Can we test this? As in, let's check that the current job does not produce an artifact; then perhaps @ssiccha can push a commit which introduces a failure somewhere to see what happens?

Comment (ssiccha) [reply .github/workflows/release.yml:51]: This is such a niice feature. This did not upload an artifact: 
https://github.com/gap-system/gap/pull/4307/checks?check_run_id=2021086422 
I'll now push something that fails.

Comment (ssiccha) [reply .github/workflows/release.yml:53]: This works. Here I introduced an error and it is currently uploading artifacts:
https://github.com/gap-system/gap/pull/4307/checks?check_run_id=2021529089

I'm removing my error commit again

**Review by ssiccha — APPROVED** [review]
LGTM

**Pull request merged**